### ISSUE TITLE
unicode file names not handled properly by HydroShare

### ISSUE THREAD

**Issue by alvacouch** [hydroshare/hydroshare#3024]
@martinseul @dtarb @sblack-usu
The plot thickens as I study the Unicode problem, as manifest in cuahsi-dev-2, resource `3135543760f24fa5a89c096d67b55e7f`, directory `data/contents`, filename `Appendix H â Monitoring and Assessment of Critical Thermal Dynamics in Upper Middle Fork of the John Day River, 2008-2016.pdf`. 

The Unicode character in this filename is **â**.

The full story of this file's ingestion:
 
* Windoze created a zipfile with Unicode in it. 
* Linux actually understood the zipfile Unicode. (verification: stack overflow post: https://unix.stackexchange.com/questions/38055/utf-8-filenames) 
* iRODS actually understood the zipfile Unicode. (verification: iRODS team) 
* The file was correctly ingested into Django. 

So, what's the problem?

* All of the URLs for this file are invalid, because they contain unescaped Unicode (and a bunch of other unescaped stuff like whitespace). This includes public URLS that a user would use to download the file, as well as private URLs that invoke REST AJAX features. Thus, almost any action directed at that specific file fails, due to Unicode encoding errors in the REST request!

Thus, the real solution is to fix the URLs, not the file. This must be a systematic solution in which **all URLs referring to files or file paths get fixed,** or there will be some case that doesn't work. Affected regions of code include large file download, resource landing page, etc. URLs listed to reference thsi file in resourcemetadata.xml, resourcemap.xml, and all composite resource metadata files are also invalid.

**Comment by alvacouch** [comment]
A few background notes

1. Linux only understands Unicode in the UTF-8 format. Other formats are not understood. 
2. There is a standard escape sequence for encoding UTF-8 and a library function somewhere to do it. I need to locate that function.  
3. Running this library function on a filename not containing unicode does nothing. 
4. Thus running the function on every URL makes sense.

**Comment by dtarb** [comment]
As an additional test case, here is a zip file provided by a user a while back that was a problem.  
[Filenames.zip](https://github.com/hydroshare/hydroshare/files/2621056/Filenames.zip)
It contains several files with unicode character names.  If I attempt to unzip in hydroShare today I get a folder with only one file, i.e. several of the files are not extracted.  

Will the solution suggested resolve this.

**Comment by alvacouch** [comment]
@dtarb This was a thorny problem to figure out, because 'utf-8' is not a true type in python, it is a subtype of the python type `string`. A string can be ASCII or `utf-8` _depending upon its content_.  So, routines that take ASCII or `Unicode` presume that _all `string` values are ASCII_, and balk at special characters that are not ASCII in those `string` values.

**Comment by alvacouch** [comment]
@cuahsimarko @martinseul @sblack-usu I've completely changed my mind on how to address this: 

1. Linux supports unicode in filenames. 
2. Linux terminals support utf-8 in filenames. 
3. Linux converts ls output from unicode to utf-8 automagically. 
4. iRODS ls interfaces return utf-8 that can be converted to unicode. 

Thus the path of least resistance is to 

1. fix directory listings to report filenames in unicode rather than ASCII or utf-8, in `django_irods`. 
2. fix filename processing to completely support unicode for all operations. This means avoiding ASCII-only operations. 
3. fix the UI to html-encode unicode for UI operations. 
4. fix URLs for filenames by REST escaping them. 

Then the rest-escape version will be converted to unicode, which will match the unicode version of the ls command. 

The bottom line is that inside hydroshare, all filenames are unicode. Outside, they're converted to whatever they need to be.

**Comment by eclark-cuahsi** [comment]
Related to #2538 and #3024

Discussed that this is a complex issue that would need to be carefully worked out.

**Comment by pkdash** [comment]
@alvacouch and @dtarb The only issue that I can see is with the "Get File URL" which shows the URL not properly encoded as shown in this screenshot:

![image](https://user-images.githubusercontent.com/2523434/202554389-a57139b8-16db-4ae8-b3aa-d25e9db75a35.png)
 
The file paths are properly encoded in resourcemap.xml and aggregation map xml files.  I will be fixing url for the 'Get File URL'.
